src/python/pants/pantsd/pailgun_server.py

```diff
@@ -178,19 +196,74 @@ def handle_request(self):
     with self.lifecycle_lock():
       self._handle_request_noblock()
 
+  def _should_poll_forever(self, timeout):
+    return timeout < 0
+
+  def _should_keep_polling(self, timeout, time_polled):
+    return self._should_poll_forever(timeout) or time_polled < timeout
+
+  @contextmanager
+  def ensure_request_is_exclusive(self, environment, request):
+    """
+    Ensure that this is the only pants running.
+
+    We currently don't allow parallel pants runs, so this function blocks a request thread until
+    there are no more requests being handled.
+    """
+    # TODO add `did_poll` to pantsd metrics
+
+    timeout = float(environment.pop('PANTSD_REQUEST_TIMEOUT_LIMIT'))
+
+    @contextmanager
+    def yield_and_release(time_waited):
+      try:
+        self.logger.debug("request lock acquired {}".format("on the first try" if time_waited == 0 else "in {}s".format(time_waited)))
+        yield
+      finally:
+        self.handling_request_lock.release()
+        self.logger.debug("released request lock")
+
+    time_polled = 0.0
+    polling_interval = 1.0 # check the lock every second
+    self.logger.debug("request {} is trying to aquire the request lock".format(request))
+
+    # Optimistically try to acquire the lock without blocking, in case we are the only request being handled.
+    # This could be merged into the while below, but separating this special case for logging helps.
+    if self.handling_request_lock.acquire(blocking=False):
+      with yield_and_release(time_polled):
+        yield
+    else:
+      # We have to wait for another request to finish being handled.
+      NailgunProtocol.send_stderr(request, "Another pants run was found, starting waiting {} for it to finish\n".format(
+        "forever" if self._should_poll_forever(timeout) else "for up to {}s".format(timeout)
+      ))
+      while not self.handling_request_lock.acquire(blocking=True, timeout=polling_interval):
+        time_polled += polling_interval
+        if self._should_keep_polling(timeout, time_polled):
+          NailgunProtocol.send_stderr(request, "Waiting for request to finish ({}s total)...\n".format(time_polled))
+        else: # We have timed out.
+          self.logger.error("Request {} timed out while waiting for another pants run to finish".format(request))
```

Comment:
If you end up calling it a "pants command" instead of "pants run", then you'll need to update multiple references around here.

I'm not sure which is better, both in terms of correctness and understanding for users. Command makes more sense to me, but also a fan of run. Maybe "invocation"?

Reply:
"invocation" seems perfect here, thanks!

Reply:
Although here, for logging, we really want to use `"Request"`, since a request is an actual concept in this class (a Pailgun Request).

Reply:
Sorry, the word "run" appears in the same line and I hadn't noticed. My bad, scratch that.